File: backend/app/schemas/campaign.py
"""
Pydantic schemas for the persistence layer (campaigns, candidates, notes,
conversations, activity). These mirror the Supabase tables in migration 0001
and provide strong typing at the API boundary.

The heavy AI output shapes (CandidateResult, CopilotResponse) are NOT redefined
here — they already live in schemas.batch / schemas.copilot and are stored
verbatim as jsonb. This module only models the persistence metadata around them.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.batch import RankingWeights


class CampaignStatus(str, Enum):
    draft = "draft"
    active = "active"
    paused = "paused"
    archived = "archived"


class PipelineStage(str, Enum):
    sourced = "sourced"
    screening = "screening"
    shortlisted = "shortlisted"
    interview = "interview"
    offer = "offer"
    hired = "hired"
    rejected = "rejected"


# ── Campaigns ────────────────────────────────────────────────────────────────
class CampaignCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    company: Optional[str] = Field(default=None, max_length=200)
    role_title: Optional[str] = Field(default=None, max_length=200)
    department: Optional[str] = None
    location: Optional[str] = None
    employment_type: Optional[str] = None
    job_description: str = ""
    ranking_weights: Optional[RankingWeights] = None
    status: CampaignStatus = CampaignStatus.draft


class CampaignUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    company: Optional[str] = None
    role_title: Optional[str] = None
    department: Optional[str] = None
    location: Optional[str] = None
    employment_type: Optional[str] = None
    job_description: Optional[str] = None
    ranking_weights: Optional[RankingWeights] = None
    status: Optional[CampaignStatus] = None


class Campaign(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    recruiter_id: str
    title: str
    company: Optional[str] = None  # stored in metadata.company (no dedicated column)
    role_title: Optional[str] = None
    department: Optional[str] = None
    location: Optional[str] = None
    employment_type: Optional[str] = None
    job_description: str = ""
    jd_storage_path: Optional[str] = None
    ranking_weights: dict[str, Any] = Field(default_factory=dict)
    status: CampaignStatus = CampaignStatus.draft
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Aggregates hydrated on demand (not columns).
    candidate_count: Optional[int] = None
    total_candidates: Optional[int] = None
    awaiting_analysis: Optional[int] = None
    average_match_score: Optional[float] = None
    last_activity_at: Optional[datetime] = None


# ── Candidates ───────────────────────────────────────────────────────────────
class Candidate(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    campaign_id: str
    recruiter_id: str
    full_name: str = ""
    email: Optional[str] = None
    phone: Optional[str] = None
    resume_path: Optional[str] = None
    resume_filename: Optional[str] = None
    source_batch_id: Optional[str] = None
    stage: PipelineStage = PipelineStage.sourced
    is_favorite: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Latest analysis, hydrated on demand.
    latest_analysis: Optional[dict[str, Any]] = None


class CandidateStageUpdate(BaseModel):
    stage: PipelineStage


# ── Notes ────────────────────────────────────────────────────────────────────
class NoteCreate(BaseModel):
    body: str = Field(..., min_length=1)
    pinned: bool = False


class RecruiterNote(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    candidate_id: str
    campaign_id: str
    recruiter_id: str
    body: str
    pinned: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ── Copilot conversations ────────────────────────────────────────────────────
class ConversationCreate(BaseModel):
    candidate_id: str
    title: str = "New conversation"


class CopilotConversation(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    candidate_id: str
    campaign_id: str
    recruiter_id: str
    title: str = "New conversation"
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class CopilotMessageRecord(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    conversation_id: str
    recruiter_id: str
    role: str
    content: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None


# ── Persistence request wrappers ─────────────────────────────────────────────
class PersistBatchRequest(BaseModel):
    """Attach a completed batch analysis to a campaign for storage."""

    campaign_id: str


class BulkCandidateIds(BaseModel):
    """Payload for bulk candidate actions (delete, etc.)."""

    candidate_ids: list[str] = Field(default_factory=list)


class ActivityEvent(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    recruiter_id: str
    campaign_id: Optional[str] = None
    candidate_id: Optional[str] = None
    type: str
    summary: str = ""
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None

File: frontend/types/campaign.ts
/** Types mirroring the backend persistence schemas (schemas/campaign.py). */

export type CampaignStatus = 'draft' | 'active' | 'paused' | 'archived';

export type PipelineStage =
  | 'sourced'
  | 'screening'
  | 'shortlisted'
  | 'interview'
  | 'offer'
  | 'hired'
  | 'rejected';

export interface RankingWeights {
  skills: number;
  experience: number;
  projects: number;
  ats: number;
  education: number;
  semantic: number;
  achievements: number;
}

export interface Campaign {
  id: string;
  recruiter_id: string;
  title: string;
  company?: string | null;
  role_title?: string | null;
  department?: string | null;
  location?: string | null;
  employment_type?: string | null;
  job_description: string;
  jd_storage_path?: string | null;
  ranking_weights: Record<string, number>;
  status: CampaignStatus;
  metadata: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
  candidate_count?: number | null;
  // Dashboard aggregates (hydrated by GET /campaigns).
  total_candidates?: number | null;
  awaiting_analysis?: number | null;
  average_match_score?: number | null;
  last_activity_at?: string | null;
}

export interface CampaignCreateInput {
  title: string;
  company?: string;
  role_title?: string;
  department?: string;
  location?: string;
  employment_type?: string;
  job_description?: string;
  status?: CampaignStatus;
  ranking_weights?: RankingWeights;
}

export type CampaignUpdateInput = Partial<CampaignCreateInput>;

export interface Candidate {
  id: string;
  campaign_id: string;
  recruiter_id: string;
  full_name: string;
  email?: string | null;
  phone?: string | null;
  resume_path?: string | null;
  resume_filename?: string | null;
  source_batch_id?: string | null;
  stage: PipelineStage;
  is_favorite: boolean;
  metadata: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
  latest_analysis?: Record<string, unknown> | null;
}

export interface RecruiterNote {
  id: string;
  candidate_id: string;
  campaign_id: string;
  recruiter_id: string;
  body: string;
  pinned: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface RecruiterProfile {
  id: string;
  email: string;
  full_name?: string | null;
  company?: string | null;
  job_title?: string | null;
  avatar_url?: string | null;
  onboarded: boolean;
}

export interface ActivityEvent {
  id: string;
  recruiter_id: string;
  campaign_id?: string | null;
  candidate_id?: string | null;
  type: string;
  summary: string;
  payload: Record<string, unknown>;
  created_at?: string;
}
